### DISCUSSION TITLE
Pandapower Diagnostic Tool Flags Correct Impedance Values as Too Low and Incorrect Transformer Loading Calculation

### DISCUSSION THREAD

**Discussion by johannanana** [e2nIEE/pandapower#2376] [Q&A]
I’m currently working on a power flow analysis using Pandapower for a small distribution network. The network consists of 16 buses, several lines, and a transformer connecting the network to a medium-voltage grid. Each line’s impedance is manually set based on known parameters. As input I give my pandapower python script a excel file, the network is set up correctly (the graph looks good, everything runs)

However, I’ve encountered a couple of significant issues that I can't seem to resolve:

Diagnostic Tool Flags Correct Impedance Values as Too Low: 
Input Data: I manually input the impedance values for the lines, with resistance (r_ohm_per_km) set to 0.322097 ohms/km and reactance (x_ohm_per_km) set to 0.073199 ohms/km. These values are consistent and correct according to my data sources.
Issue: Despite this, when I run Pandapower's diagnostic tool, it consistently flags several lines as having impedance values close to zero (r_ohm <= 0.001 or x_ohm <= 0.001). This seems to be a misinterpretation, as the values are clearly above the threshold.

Attempts to Resolve: I’ve checked the values immediately after importing them, and they are correctly set in the Pandapower model. The issue somehow only appears while running the diagnostic tool (and maybe the load flow calculation). Additionally, after running the power flow calculation, the values remain correct, but the diagnostic tool still reports the same problem.

Incorrect Transformer Loading Calculation: 
Input Data: The network includes a transformer rated at 0.25 MVA (20/0.4 kV). I’ve set up loads across multiple buses and expected the transformer loading to vary significantly with changes in load.
Issue: No matter how I change the loads (e.g., increasing each house connection load from 0.01 MW to 0.1 MW), the transformer loading percentage remains nearly constant and unrealistically low. This is highly unexpected and suggests that the power flow calculation is not reflecting the actual network conditions. I also changed the type of transformer and as that changed the loading percent it stays unrealistically and when staying with the same type of transformer the changing load changes almost nothing.

I’ve used both the old (pp.runpp()) and new (pp.run.runpp_pgm()) power flow calculation functions in Pandapower, with similar results.

I’m looking for advice on how to resolve these issues. Has anyone encountered similar problems with Pandapower misinterpreting impedance values or calculating transformer loading incorrectly? Any suggestions for troubleshooting or alternative methods to ensure accurate power flow analysis would be greatly appreciated. I have also problems finding a log file so I could check where exactly the errors occur.

here there are the values of the transformers compared: 
![image](https://github.com/user-attachments/assets/697353f7-dfd3-4184-9fa9-9c0a603afbd8)

**Comment by johannanana** [comment]
New findings/ theories to be validated by someone :) : 

- could it be that pandapower sets every cable length that is under 15m to zero? -> with all cable lengths longer the impedance problem vanishes. 

- transformer problem is more complicated. I traced it back to the pandapower_converter.py file where there is a function "create_pgm_input_transformers", which does something with the transformer. the problematic calculation is "rx_mag = mag_g / np.sqrt(i_no_load * i_no_load * 1e-4 - mag_g * mag_g)" in line 706. I am not sure what rx_mag stands for. But i_no_load is 0.32 and mag_g is 0.0032 and therefore the squareroot is 0 which concludes to a division by zero. 

How can I fix this problem, I use a standard type transformer from pandapower and my problem occurs with every transformer I tried. 
Can someone explain this to me? 

Thank you!

**Comment by johannanana** [comment]
new findings II: 

- in the Slack-Sheet of the input-excel the wrong bus was connected (has to be the medium voltage grid bus aka the external grid representative) -> so the big trafo-problem is solved

-  not solved is this error:

```python 
pandapower_converter.py:712: RuntimeWarning: divide by zero encountered in divide
  rx_mag = mag_g / np.sqrt(i_no_load * i_no_load * 1e-4 - mag_g * mag_g)
  ```
  
  I tried to set up a new pandapower file with a small test grid from the pandapower website and then the transformer loading values are changing realistically, but the rx_mag error is still on. I printed i_no_load and mag_g and those are 0.32 and 0.0032 which are 0 when connected through the formula. From the pandapower_converter.py file I know that the rx_mag is "only" used to check the transformer values, and not in further calculations, but it is unnerving either way.

Here the code how the value is used:

```python
# Calculate rx ratio of magnetising branch
mag_g = np.divide(pfe, sn_mva * 1000, where=valid)
mag_g[np.logical_not(valid)] = np.nan


 rx_mag = mag_g / np.sqrt(i_no_load * i_no_load * 1e-4 - mag_g * mag_g)
 # positive and zero sequence magnetising impedance must be equal.
# mag0_percent = z0mag / z0.
checks = {
            "vk0_percent": np.allclose(vk_percent, vk0_percent) or np.isnan(vk0_percent).all(),
            "vkr0_percent": np.allclose(vkr_percent, vkr0_percent) or np.isnan(vkr0_percent).all(),
            "mag0_percent": np.allclose(i_no_load * 1e-2, 1e4 / (vk0_percent * mag0_percent))
            or np.isnan(mag0_percent).all(),
            "mag0_rx": np.allclose(rx_mag, mag0_rx) or np.isnan(mag0_rx).all(),
            "si0_hv_partial": np.isnan(
                self._get_pp_attr("trafo", "si0_hv_partial", expected_type="f8", default=np.nan)
            ).all(),
        }
        if not all(checks.values()):
            failed_checks = ", ".join([key for key, value in checks.items() if not value])
            logger.warning(f"Zero sequence parameters given in trafo shall be ignored:{failed_checks}")

``` 

- also not yet understood is the impedance problem, just changing the length of cables doesn't seem like the correct approach.  So I am searching for another way of securing that the impedance doesn't go to zero. 
Any ideas what might be behind that?

**Comment by vogt31337** [comment]
@johannanana , without the actual grid, it is kinda hard to track your actual problem. Short lines are a problem in powerflow calculation. This is not a pandapower problem. so If the line values are given in per km and you enter 0.07 ohms/km * 0.015km the value will be correctly flagged as really low, since this can lead to mathematical instabilities. For the transformer part, please open an issue and post a minimal example showing the problem.
